cpp/src/arrow/util/async_generator.h

```diff
@@ -785,6 +786,24 @@ class ReadaheadGenerator {
 template <typename T>
 class PushGenerator {
   struct State {
+    explicit State(util::BackpressureOptions backpressure)
+        : backpressure(std::move(backpressure)) {}
+
+    void OpenBackpressureIfFree() {
+      if (backpressure.toggle && !backpressure.toggle->IsOpen() &&
+          result_q.size() < backpressure.resume_if_below) {
+        backpressure.toggle->Open();
+      }
+    }
+
+    void CloseBackpressureIfFull() {
+      if (backpressure.toggle && backpressure.toggle->IsOpen() &&
+          result_q.size() > backpressure.pause_if_above) {
+        backpressure.toggle->Close();
```

Comment:
Is there any value checking IsOpen since Close performs the same check internally?

Reply:
No.  I removed these checks.